### PR TITLE
tie networking and ratelimit using annotations

### PR DESCRIPTION
**Pull request by rahulanand16nov**
This PR:
- Move networking and ratelimit mapping out of API to annotations in RateLimitPolicy.
- Fixes `BOTH` stage value not turning into two ratelimit fields
- Makes uniqueness requirement on the route names in RateLimitPolicy.
- Adds annotation to the virtualservice when RateLimitPolicy is created to map back to it.
- Removes selectors from API resource.

## Steps to verify functionality
1. `kind create cluster`
2. `kuadrantctl install`
3. `kubectl apply -f examples/toystore/toystore.yaml`
4. `kubectl apply -f examples/toystore/virtualservice.yaml`
5. `make install`
6. `make run`
7. `kubectl apply -f config/samples/apim_v1alpha1_ratelimitpolicy.yaml`
8. `kubectl port-forward -n kuadrant-system svc/kuadrant-gateway 9080:80`
9. `curl -X POST -H "Host: toy.toystore.com" localhost:9080/admin/addToy -v` 2 times

### CODE REVIEW

File: config/samples/apim_v1alpha1_ratelimitpolicy.yaml
Comment (maleck13) [inline config/samples/apim_v1alpha1_ratelimitpolicy.yaml:8]: I think this should be part of the spec as per our discussion today 

```
targetRef:
   type: VirtualService
   name: myservice
```

Comment (rahulanand16nov) [reply config/samples/apim_v1alpha1_ratelimitpolicy.yaml]: Updated. Just instead of `targetRef`, I have chosen `networkingRef`

File: controllers/apim/ratelimitpolicy_controller.go
Comment (maleck13) [inline controllers/apim/ratelimitpolicy_controller.go]: Does this need to be removed?

Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go]: I guess that's the old commit, I cannot find that in the new commit.

Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go]: Okay, that constant was still present, removed.

File: controllers/apim/ratelimitpolicy_controller.go
Comment (eguzki) [inline controllers/apim/ratelimitpolicy_controller.go:206]: Objects needs to be reconciled, not only created. Otherwise changes in the RLP will not update the envoyfilter objects.

Anyway, it is good enough for now

Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go:206]: Yes, just waiting for the green light before i tighten those loose screws.

Reconciling EnvoyFilter is a bit tricky since need to be aware of VS/HTTPR change as well.

File: controllers/apim/ratelimitpolicy_controller.go
Comment (eguzki) [inline controllers/apim/ratelimitpolicy_controller.go:373]: This should not be done. Structured data should not be serialized to apply string substitution and then de-serialize 
again. 

You can build the unstructured (based on maps and interfaces) actions list object and then apply to the `patchUnstructured` object.


For now it is ok

File: controllers/apim/ratelimitpolicy_controller.go
Comment (eguzki) [inline controllers/apim/ratelimitpolicy_controller.go:144]: There needs to be a good reason to update a resource that the controller is not owner of. 

What is the purpose of adding that annotation?

File: controllers/apim/ratelimitpolicy_controller.go
Comment (eguzki) [inline controllers/apim/ratelimitpolicy_controller.go:152]: envoyfilters reconciliation in a method called "reconcileWithVirtualService` feels weird

File: controllers/apim/ratelimitpolicy_controller.go
Comment (eguzki) [inline controllers/apim/ratelimitpolicy_controller.go:335]: I wonder if the envoyfilter resource name is linked to the rate limiting policy, more than  to the gateway

**Review by eguzki — CHANGES_REQUESTED** [review]
Related to the ratelimitpolicy API, Currently it is defined with `routes.name`. Is this a better option that based on rules (like authorization policy)?

In this [gist](https://gist.github.com/eguzki/b9cabed957760fc03142dc9d85419ba5), there are rules like:
```yaml
spec:
  phases: ["postAuth"]
  rules:  # All conditions inside a single match block have AND semantics while match blocks have OR semantics
    - hosts:  # optional
      - toystore.example.com
      methods:  # optional
      - POST
      paths:  # optional
      - "/admin/addToy"
      notPaths: []  # optional
      notHosts: []  # optional
      notMethods: []  # optional
```

Obviously it is more complicated to "match" rules with the VS routes. However it is more aligned to reusable istio policies. what are your thoughts on this?

**Comment by maleck13** [comment]
> Related to the ratelimitpolicy API, Currently it is defined with `routes.name`. Is this a better option that based on rules (like authorization policy)?
> 
> In this [gist](https://gist.github.com/eguzki/b9cabed957760fc03142dc9d85419ba5), there are rules like:
> 
> ```yaml
> spec:
>   phases: ["postAuth"]
>   rules:  # All conditions inside a single match block have AND semantics while match blocks have OR semantics
>     - hosts:  # optional
>       - toystore.example.com
>       methods:  # optional
>       - POST
>       paths:  # optional
>       - "/admin/addToy"
>       notPaths: []  # optional
>       notHosts: []  # optional
>       notMethods: []  # optional
> ```
> 
> Obviously it is more complicated to "match" rules with the VS routes. However it is more aligned to reusable istio policies. what are your thoughts on this?

Currently the routematch in the envoy filter object only supports ```name``` and ```action``` so I think for this to work with the API you are proposing we would have to go the route of WASM?

**Comment by maleck13** [comment]
> Related to the ratelimitpolicy API, Currently it is defined with `routes.name`. Is this a better option that based on rules (like authorization policy)?
> 
> In this [gist](https://gist.github.com/eguzki/b9cabed957760fc03142dc9d85419ba5), there are rules like:
> 
> ```yaml
> spec:
>   phases: ["postAuth"]
>   rules:  # All conditions inside a single match block have AND semantics while match blocks have OR semantics
>     - hosts:  # optional
>       - toystore.example.com
>       methods:  # optional
>       - POST
>       paths:  # optional
>       - "/admin/addToy"
>       notPaths: []  # optional
>       notHosts: []  # optional
>       notMethods: []  # optional
> ```
> 
> Obviously it is more complicated to "match" rules with the VS routes. However it is more aligned to reusable istio policies. what are your thoughts on this?

Currently the routematch in the envoy filter object only supports name and action so I think for this to work with the API you are proposing we would have to go the route of WASM?

File: controllers/apim/ratelimitpolicy_controller.go
Comment (maleck13) [reply controllers/apim/ratelimitpolicy_controller.go:373]: perhaps we should create a TODO or issue to tack the need to change this?

File: controllers/apim/ratelimitpolicy_controller.go
Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go:310]: Not aware of that patch and difference doesn't matter that match except `clusterEnvoyPatch` adds a new cluster but the one in the PR changes the actual name to `rate_limit_cluster`. Will reuse `clusterEnvoyPatch`.

File: controllers/apim/ratelimitpolicy_controller.go
Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go:183]: same [reply](https://github.com/Kuadrant/kuadrant-controller/pull/76/files#r789825831)

File: controllers/apim/ratelimitpolicy_controller.go
Comment (maleck13) [reply controllers/apim/ratelimitpolicy_controller.go:144]: Yes not sure this is needed now that we have changed the spec to have a targetRef. Currently the rate limit policy looks in the same namespace. So we should be able to put a watch on VirtualServices and HTTPRoutes and when one changes in a way we need to do something about, look up any RLP in same namespace and identify any that target the changed resource.

File: controllers/apim/ratelimitpolicy_controller.go
Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go:373]: For some reason, deserializing actions into that unstructured patch was working properly but I agree this needs a better way. Will create an issue when the PR is merged.

File: controllers/apim/ratelimitpolicy_controller.go
Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go:144]: It provided a mechanism to map to a particular RLP when VirtualService changes and is being watched. Considering what Craig said, this can be removed and I'll update the PR.

File: controllers/apim/ratelimitpolicy_controller.go
Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go:152]: The idea is we have two networking sources, namely, VirtualService and HTTPRoute. So, when VS changes, use `reconcileWithVirtualService` and so forth for HTTPRoute. Reconciling of envoyfilters is required since they rely on gateway names from networking sources.

**Comment by maleck13** [comment]
tried this out locally. Added new rate limit rule for the get descriptor all worked as expected. I think when we demo this it would be good to show 

1) A global rate limit for all end points
2) a non authenticated rate limit for an endpoint
3) an authenticated rate limit for the admin endpoint

WDYT?

File: controllers/apim/ratelimitpolicy_controller.go
Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go:335]: ratelimit filters patch is required once per gateway so if it's per RLP then we'll have redundant patches to the same gateway if multiple RLPs target same gateway.

File: controllers/apim/ratelimitpolicy_controller.go
Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go:152]: Yes, it feels a bit weird to me as well. I'll change this as well :thinking:

**Comment by rahulanand16nov** [comment]
I think, we can move the route patches into a single EnvoyFilter resource and hence management of them. Will try that as well :)

File: controllers/apim/ratelimitpolicy_controller.go
Comment (maleck13) [reply controllers/apim/ratelimitpolicy_controller.go:144]: Do we want to remove this as part of this PR?

**Comment by maleck13** [comment]
@eguzki are you happy with the PR in its current state ?
@rahulanand16nov I think we should capture any outstanding items in Jira that we don't intend to fix here so that we do not lose track of them. This could be a simple ```Harden RateLimitPolicy reconciler ``` jira with a set of bullet points need to complete the work

**Comment by rahulanand16nov** [comment]
> @eguzki are you happy with the PR in its current state ? @rahulanand16nov I think we should capture any outstanding items in Jira that we don't intend to fix here so that we do not lose track of them. This could be a simple `Harden RateLimitPolicy reconciler ` jira with a set of bullet points need to complete the work

OOO for today, I have some changes planned that I'll push tomorrow and add rest to a JIRA. Let's wait till tomorrow and we can merge this PR and move focus to auth.

**Comment by rahulanand16nov** [comment]
I'll merge this for now and move the remaining details into the existing spike JIRA. Moving the focus to auth.

File: controllers/apim/ratelimitpolicy_controller.go
Comment (rahulanand16nov) [reply controllers/apim/ratelimitpolicy_controller.go:145]: Removing this and watching VS/HR is a relatively big change. I'll add them in the next PR.

**Pull request merged**
